### PR TITLE
refactor: event types infinite scroll

### LINKED ISSUE #16023: [CAL-4092] Event type infinite scrolling
<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-4092](https://linear.app/calcom/issue/CAL-4092/event-type-infinite-scrolling)</sub>

### PR DESCRIPTION
**Pull request by Udit-takkar**
## What does this PR do?

- Fixes https://github.com/calcom/cal.com/issues/16023
- Fixes CAL-4092 (Linear issue number - should be visible at the bottom of the GitHub issue description)

AFTER:-

https://github.com/user-attachments/assets/6782bd3b-0e22-4ee0-8af1-feb4d13f305c

TODO
- ~Fix type errors~

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] N/A I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com). If N/A, write N/A here and check the checkbox.
- [x] N/A I confirm automated tests are in place that prove my fix is effective or that my feature works.

### CODE REVIEW

File: apps/web/modules/event-types/views/event-types-listing-view.tsx
Comment (Udit-takkar) [inline apps/web/modules/event-types/views/event-types-listing-view.tsx:867]: This would fetch all the `eventTypeGroups`

File: packages/features/kbar/Kbar.tsx
Comment (Udit-takkar) [inline packages/features/kbar/Kbar.tsx:40]: infinte query to load all event types

File: apps/web/modules/event-types/views/event-types-listing-view.tsx
Comment (Udit-takkar) [inline apps/web/modules/event-types/views/event-types-listing-view.tsx:275]: had to change this logic to support pagination

File: apps/web/modules/event-types/views/event-types-listing-view.tsx
Comment (Udit-takkar) [inline apps/web/modules/event-types/views/event-types-listing-view.tsx:841]: now we only display team tabs

File: packages/trpc/server/routers/viewer/eventTypes/getEventTypesFromGroup.handler.ts
Comment (hariombalhara) [inline packages/trpc/server/routers/viewer/eventTypes/getEventTypesFromGroup.handler.ts:53]:
```suggestion
            ...(!!filters?.schedulingTypes
```

should be sufficient

File: packages/trpc/server/routers/viewer/eventTypes/getEventTypesFromGroup.handler.ts
Comment (hariombalhara) [inline packages/trpc/server/routers/viewer/eventTypes/getEventTypesFromGroup.handler.ts:50]: I think it is done now

File: packages/trpc/server/routers/viewer/eventTypes/getEventTypesFromGroup.handler.ts
Comment (hariombalhara) [inline packages/trpc/server/routers/viewer/eventTypes/getEventTypesFromGroup.handler.ts:55]: I think `filters.schedulingTypes` isn't being passed yet. It is supposed to be passed from UserTable.

We could add this when we support that component as well with it.

File: apps/web/modules/event-types/views/event-types-listing-view.tsx
Comment (hariombalhara) [inline apps/web/modules/event-types/views/event-types-listing-view.tsx:1]: These are huge changes. How about we have a separate component for it and we use the new one when a infiniteScrollEventType feature is enabled ? We already have features table.

I observe that we aren't touching getEventTypesByViewer so it seems that it is doable without many changes.

It would let us ship this change faster I believe.

File: apps/web/lib/event-types/getServerSideProps.tsx
Comment (hariombalhara) [inline apps/web/lib/event-types/getServerSideProps.tsx:15]: Could be moved below session `nullish` check. The condition would need two less ?s

File: packages/features/flags/config.ts
Comment (hariombalhara) [inline packages/features/flags/config.ts:16]: I don't think it is needed

File: packages/features/eventtypes/components/CreateEventTypeDialog.tsx
Comment (hariombalhara) [inline packages/features/eventtypes/components/CreateEventTypeDialog.tsx:121]: Shouldn't we handle it based on flag?

Comment (hariombalhara) [inline packages/features/eventtypes/components/CreateEventTypeDialog.tsx:121]: This worked. Earlier newly created event wasn't showing in the list without refresh(when the flag is disabled)
```suggestion
       if (isInfiniteScrollEnabled) {
        await utils.viewer.eventTypes.getEventTypesFromGroup.invalidate();
      } else {
        await utils.viewer.eventTypes.getByViewer.invalidate();
      }
```

File: packages/lib/server/repository/eventType.ts
Comment (hariombalhara) [inline packages/lib/server/repository/eventType.ts:211]: We could remove the duplicates at 3 places and compute cursor and take at a single place at the top.

File: packages/lib/server/repository/eventType.ts
Comment (hariombalhara) [inline packages/lib/server/repository/eventType.ts:221]:
```suggestion
    limit = 10,
```

This isn't breaking anything as it is a new function but I would prefer limit to be passed explicitly as it could cause the same problem where only 10 records are being returned and user isn;t aware of it.

File: packages/trpc/server/routers/viewer/eventTypes/getByViewer.schema.ts
Comment (hariombalhara) [inline packages/trpc/server/routers/viewer/eventTypes/getByViewer.schema.ts:27]: We should move it to it's .schema file

File: apps/web/modules/event-types/views/event-types-listing-view.tsx
Comment (hariombalhara) [inline apps/web/modules/event-types/views/event-types-listing-view.tsx:804]: nit: I think this comment could be stale. Let's not copy it here.

File: apps/web/lib/hooks/useInViewObserver.ts
Comment (hariombalhara) [inline apps/web/lib/hooks/useInViewObserver.ts:48]: It could be just a flag in `useInViewObserver` right or better we could support a root argument in there

**Comment by hariombalhara** [comment]
`Owner` tab should be shown selected as soon as I land on event-types page as it is showing owner's event-types. It works when selecting the tab later on

image

**Comment by hariombalhara** [comment]
@Udit-takkar This one is an important to fix bug. Event types list showing very few event-types after an event is created.
https://www.loom.com/share/37312041544847dc87e8a679c2f65049

**Comment by hariombalhara** [comment]
Another critical one. Managed event isn't showing up.
https://www.loom.com/share/c079ae77dcc44ffdb8bc6ef7358fbaac

**Comment by hariombalhara** [comment]
Sometimes Ctrl+K doesn't show event-types list but it seems to happen in main as well for me. So, we can fix it later.

**Comment by Udit-takkar** [comment]
> This one is an important to fix bug. Event types list showing very few event-types after an event is created.

Fixed

https://github.com/user-attachments/assets/3f6baf10-51f7-4cb4-b850-736c0fae759c

**Review by alishaz-polymath — DISMISSED** [review]
Code LGTM and tested the change, great improvement in performance 👏 
Will wait for @hariombalhara to finish his review before we merge this 🚀 🚀

**Review by hariombalhara — CHANGES_REQUESTED** [review]
- [x] Blocker: When I delete an event-type, it ends up showing that a lot more even-types have been deleted. Looks like it removes an entire page from UI. In reality it works fine, it fixes after refresh.
https://www.loom.com/share/0967d6ce6e064632bd6db3f51f3ce86a

- [x] Non Blocker(can be fixed in followup):  When entire list is loaded, at that moment creating an event-type still resets the entire list and shows only the last page
https://www.loom.com/share/03abc2b6a423495b9c0c1c641a980959

Found the above bugs. The blocker one I think should be fixed before the PR can be merged.

**Comment by hariombalhara** [comment]
With feature disabled, after duplication I don't see the newly created event-type(without refresh)

https://www.loom.com/share/46a94ff6c91d4c2caacb63c06dd56ab4

**Comment by Udit-takkar** [comment]
> With feature disabled, after duplication I don't see the newly created event-type(without refresh)

I tried using `setInfiniteData` https://trpc.io/docs/client/react/useInfiniteQuery#setinfinitedata to update cache but it isn't working. I can use `await utils.viewer.eventTypes.invalidate()` but that would reset the event type list and display the next list.

**Comment by hariombalhara** [comment]
> > With feature disabled, after duplication I don't see the newly created event-type(without refresh)
> 
> I tried using `setInfiniteData` [trpc.io/docs/client/react/useInfiniteQuery#setinfinitedata](https://trpc.io/docs/client/react/useInfiniteQuery#setinfinitedata) to update cache but it isn't working. I can use `await utils.viewer.eventTypes.invalidate()` but that would reset the event type list and display the next page.

But this is for feature disabled and there is no infinite query in there right?

**Review by hariombalhara — DISMISSED** [review]
Let's ship it and then improve/fix the bugs. I don't see a major blocker.

Great work @Udit-takkar as always !!

**Comment by hariombalhara** [comment]
> > > With feature disabled, after duplication I don't see the newly created event-type(without refresh)
> > 
> > 
> > I tried using `setInfiniteData` [trpc.io/docs/client/react/useInfiniteQuery#setinfinitedata](https://trpc.io/docs/client/react/useInfiniteQuery#setinfinitedata) to update cache but it isn't working. I can use `await utils.viewer.eventTypes.invalidate()` but that would reset the event type list and display the next page.
> 
> But this is for feature disabled and there is no infinite query in there right?

Fixed it here, it was a typo [`abf2b93` (#15949)](https://github.com/calcom/cal.com/pull/15949/commits/abf2b935fa18eb9cb70f43fcd3143a8d54fa36fd)

**Review by hariombalhara — DISMISSED** [review]

**Comment by hariombalhara** [comment]
Also, the following is still not fixed. This is a weird issue with infiniteQuery, we have to manually update the infiniteData with the newly added event-type(which requires the create return value to be of the same format as eventType listing).

Let's fix it in a followup

>  Non Blocker(can be fixed in followup): When entire list is loaded, at that moment creating an event-type still resets the entire list and shows only the last page

**Review by hariombalhara — DISMISSED** [review]

File: apps/web/lib/hooks/useInViewObserver.ts
Comment (hariombalhara) [inline apps/web/lib/hooks/useInViewObserver.ts:26]: @Udit-takkar It is the fix for failing build as document was being used during SSR earlier

**Review by alishaz-polymath — APPROVED** [review]

**Pull request merged**
